src/include/sof/audio/smart_amp/smart_amp.h

```diff
@@ -11,180 +11,300 @@
-};
+typedef void (*smart_amp_src_func)(struct smart_amp_mod_stream *src_mod,
+				   uint32_t frames,
+				   const struct audio_stream __sparse_cache *src,
+				   const int8_t *chan_map);
 
-struct smart_amp_param_struct_t {
-	struct param_buf_struct_t param;	/* variable to keep last parameter ID/value */
-	struct smart_amp_caldata caldata;	/* model data buffer */
-	int pos;				/* data position for read/write */
-	int max_param;				/* keep max number of DSM parameters */
-};
+typedef void (*smart_amp_sink_func)(const struct smart_amp_mod_stream *sink_mod,
+				    uint32_t frames,
+				    const struct audio_stream __sparse_cache *sink);
 
-struct smart_amp_mod_struct_t {
-	struct smart_amp_buf_struct_t buf;
-	void *dsmhandle;
-	/* DSM variables for the initialization */
-	int delayedsamples[SMART_AMP_FF_MAX_CH_NUM << 2];
-	int circularbuffersize[SMART_AMP_FF_MAX_CH_NUM << 2];
-	/* Number of samples of feed forward and feedback frame */
-	int ff_fr_sz_samples;
-	int fb_fr_sz_samples;
-	int channelmask;
-	/* Number of channels of DSM */
-	int nchannels;
-	/* Number of samples of feed forward channel */
-	int ifsamples;
-	/* Number of samples of feedback channel */
-	int ibsamples;
-	/* Number of processed samples */
-	int ofsamples;
-	/* Channel bit dempth */
-	int bitwidth;
-	struct smart_amp_param_struct_t param;
+struct smart_amp_func_map {
+	uint16_t comp_fmt;
+	uint16_t mod_fmt;
+	smart_amp_src_func src_func;
+	smart_amp_sink_func sink_func;
 };
 
-typedef void (*smart_amp_func)(const struct comp_dev *dev,
-			       const struct audio_stream __sparse_cache *source,
-			       const struct audio_stream __sparse_cache *sink,
-			       const struct audio_stream __sparse_cache *feedback,
-			       uint32_t frames);
-struct smart_amp_func_map {
-	uint16_t frame_fmt;
-	smart_amp_func func;
+extern const struct smart_amp_func_map src_sink_func_map[];
+
+smart_amp_src_func smart_amp_get_src_func(uint16_t comp_fmt, uint16_t mod_fmt);
+smart_amp_sink_func smart_amp_get_sink_func(uint16_t comp_fmt, uint16_t mod_fmt);
+
+/******************************************************************************
+ * Inner model operations (mod ops):                                          *
+ *    Model implementations are mutual exclusive (separated by Kconfig). It   *
+ *    must be only one solution applicable per build. The solution-specific   *
+ *    implementation is placed in its own C file smart_amp_${SOLUTION}.c      *
+ ******************************************************************************/
+
+/**
+ * Creates the self-declared model data struct.
+ * Refer to "struct smart_amp_mod_data_base" for details of declaration.
+ * Args:
+ *    dev - the comp_dev object for logger tracing use only.
+ * Returns:
+ *    The pointer of "smart_amp_mod_data_base" which is the leading member in
+ *    the created model data struct.
+ */
+struct smart_amp_mod_data_base *mod_data_create(const struct comp_dev *dev);
+
+struct inner_model_ops {
```

Comment:
I assume all ops are mandatory ? We probably want to say in comments so developers know if any are optional.

Reply:
Yes they are all mandatory. I will add to the comment.